lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -146,97 +145,117 @@ public void OnNext(IAllocatedEvaluator allocatedEvaluator)
         }
 
         /// <summary>
-        /// Specfies the Map or Update task to run on the active context
+        /// Specifies the Map or Update task to run on the active context
         /// </summary>
-        /// <param name="activeContext"></param>
-        public void OnNext(IActiveContext activeContext)
+        /// <param name="value"></param>
+        public void OnNext(IActiveContext value)
         {
-            Logger.Log(Level.Verbose, string.Format("Received Active Context {0}", activeContext.Id));
+            Logger.Log(Level.Verbose, string.Format("Received Active Context {0}", value.Id));
 
-            if (_groupCommDriver.IsMasterTaskContext(activeContext))
+            if (_groupCommDriver.IsMasterTaskContext(value))
             {
                 _reachedUpdateTaskActiveContext = true;
                 RequestMapEvaluators(_dataSet.Count);
+            }
 
-                var partialTaskConf =
-                    TangFactory.GetTang()
-                        .NewConfigurationBuilder(new[]
-                        {
-                            TaskConfiguration.ConfigurationModule
-                                .Set(TaskConfiguration.Identifier,
-                                    IMRUConstants.UpdateTaskName)
-                                .Set(TaskConfiguration.Task,
-                                    GenericType<UpdateTaskHost<TMapInput, TMapOutput, TResult>>.Class)
-                                .Build(),
-                            _configurationManager.UpdateFunctionConfiguration,
-                            _configurationManager.ResultHandlerConfiguration
-                        })
-                        .BindNamedParameter(typeof(InvokeGC), _invokeGC.ToString())
-                        .Build();
+            if (_contextManager.AddContext(value))
```

Comment:
This function call now looks a little weird to me :) ... Why not have a local event trigger mechanism...In ContextManager the moment it realizes inside that it has all the contexts it generates some sort of task submission event. We can have a TaskGenerationAndSubmission class which gets triggered and start generating and submitting tasks. Driver can pass this class to ContextManager and also maintain reference to it. In future when we do fault tolerance, driver will just signal TaskGenerationAndSubmission class and it will reset itself and stop submitting tasks if it is doing so.

 Also having both ServiceAndContextConfigurationProvider<TMapInput, TMapOutput> and ContextManager seems wasteful to me.

@markusweimer any comments?